### ISSUE TITLE
NIOSSHClientUserAuthenticationDelegate Improvement Request: Result Reporting

### ISSUE THREAD

**Issue by aaronvegh** [apple/swift-nio-ssh#65]
The `NIOSSHClientUserAuthenticationDelegate` provides a single method, `nextAuthenticationRequest`, for providing offers for authenticating with a remote host. In the golden path, an auth request is successful and everyone is happy. However, in situations where the request fails, either because of incorrect credentials or an unexpected failure on the remote end, it would be useful for the delegate to receive a notification to that effect. As things stand now, the only way to know an offer fails is if the method is called again. That type of inference is something I'd rather not rely upon.

In my research of this repository, I came across `NIOSSHUserAuthenticationOutcome`; I wonder if using that here would be helpful?

**Comment by Lukasa** [comment]
Yeah, we should add more feedback to the client user auth delegate, especially as we so tightly control its execution environment.

**Comment by Joannis** [comment]
@Lukasa do you have a specific design in mind for this API? I'm happy to make a PR for it.
